onnxruntime/core/providers/trt/trt_execution_provider.cc

```diff
@@ -0,0 +1,211 @@
+    {
+        *(graph_proto.mutable_input()->Add()) = input_arg->ToProto();
+    }
+
+    for (const auto* output_arg : graph.GetOutputs())
+    {
+        *(graph_proto.mutable_output()->Add()) = output_arg->ToProto();
+    }
+
+    for (const auto* value_info : graph.GetValueInfo())
+    {
+        *(graph_proto.mutable_value_info()->Add()) = value_info->ToProto();
+    }
+
+    return graph_proto;
+}
+
+//This custom kernel is a temporary implementation.
+//TODO: change to the new function implementation.
+class ONNX_OPERATOR_KERNEL_CLASS_NAME(kTRTExecutionProvider, kMSDomain, 1, TRTKernel);
+
+ONNX_OPERATOR_KERNEL_EX(TRTKernel,
+                        kMSDomain,
+                        1,
+                        kTRTExecutionProvider,
+                        KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<float>()).SetDefaultInputsMemoryType(OrtMemTypeCPUInput).SetDefaultOutputMemoryType(OrtMemTypeCPUOutput),
+                        TRTKernel);
+
+void RegisterTRTOperatorKernels(KernelRegistry& kernel_registry) {
+  kernel_registry.Register(BuildKernelCreateInfo<ONNX_OPERATOR_KERNEL_CLASS_NAME(kTRTExecutionProvider, kMSDomain, 1, TRTKernel)>());//kOnnxDomain
+}
+
+TRTExecutionProvider::TRTExecutionProvider()
+{
+    DeviceAllocatorRegistrationInfo trt_device_info({OrtMemTypeCPU, [](int)
+    {
+        return std::make_unique<TRTPinnedAllocator>();
+    }, std::numeric_limits<size_t>::max()
+                                                    });
+    InsertAllocator(CreateAllocator(trt_device_info));
+    DeviceAllocatorRegistrationInfo default_device_info({OrtMemTypeDefault, [](int)
+    {
+        return std::make_unique<TRTAllocator>();
+    }, std::numeric_limits<size_t>::max()
+                                                        });
+    InsertAllocator(CreateAllocator(default_device_info));
+}
+
+TRTExecutionProvider::~TRTExecutionProvider() {}
+
+std::vector<std::unique_ptr<ComputeCapability>>
+        TRTExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph,
+                const std::vector<const KernelRegistry*>& /*kernel_registries*/) const
+{
+    //Construct modelproto from graph and serialize to string
+    ModelProto model_proto;
+    GraphProto graph_proto = ToGraphProto(graph);
+    model_proto.set_allocated_graph(&graph_proto);
+    string string_buf;
+    model_proto.SerializeToString(&string_buf);
+    std::vector<char> onnx_buf(string_buf.begin(), string_buf.end());
+
+    //Get supported node list
+    int verbosity = static_cast<int>(nvinfer1::ILogger::Severity::kWARNING);
+    TRTLogger trt_logger(static_cast<nvinfer1::ILogger::Severity>(verbosity));
+    const auto& trt_builder = InferObject(nvinfer1::createInferBuilder(trt_logger));
+    const auto& trt_network = InferObject(trt_builder->createNetwork());
+    const auto& trt_parser  = InferObject(nvonnxparser::createParser(trt_network.get(), trt_logger));
+
```

Comment:
Does the TRT parser process GraphProto attributes in control flow nodes, or would we have to handle finding those nested GraphProto's at this level and call the TRT parsing separately for each one?

Reply:
Not sure if I fully understand your question. We partition the graph based on GetCapability results and send the supported subgraphs to TensorRT. TensorRT will run them one by one.